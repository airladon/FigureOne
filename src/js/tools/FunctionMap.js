// @flow
// import type { Transform } from '../tools/g2';
// import { Point, getTransform, Transform } from '../tools/g2';
// import { round } from '../tools/math';
// import type { FigureElement } from './Element';
// import TimeKeeper from './webgl/TimeKeeper';

// Singleton class that contains projects global functions

class GeneralFunctionMap {
  map: { [id: string]: {
    fn: Function,
  }}

  constructor() {
    this.map = {};
  }

  add(id: string, fn: Function) {
    if (this.map == null) {
      this.map = {};
    }
    this.map[id] = {
      fn,
    };
  }

  exec(idOrFn: string | Function | null, ...args: any) {
    if (idOrFn == null) {
      return null;
    }
    if (typeof idOrFn === 'string') {
      // return this.fnMap.exec(fn, ...args);
      if (this.map[idOrFn] != null) {
        if (args.length === 0) {
          return this.map[idOrFn].fn();
        }
        return this.map[idOrFn].fn(...args);
      }
    }
    if (typeof idOrFn === 'function') {
      return idOrFn(...args);
    }
    throw new Error(`${idOrFn} needs to be a function or a string`);
  }

  execOnMaps(idOrFn: string | Function | null, mapsIn: Array<Object>, ...args: any) {
    if (idOrFn == null) {
      return null;
    }
    if (typeof idOrFn === 'string') {
      const maps = [this.map, ...mapsIn];
      for (let i = 0; i < maps.length; i += 1) {
        const map = maps[i];
        if (map[idOrFn] != null) {
          if (args.length === 0) {
            return map[idOrFn].fn();
          }
          return map[idOrFn].fn(...args);
        }
      }
      throw new Error(`FunctionMap Error: '${idOrFn}' does not exist in map`);
    }
    if (typeof idOrFn === 'function') {
      return idOrFn(...args);
    }
    throw new Error(`FunctionMap Error: '${idOrFn}' needs to be a function or a string`);
    // return null;
  }
}


class GlobalFunctionMap extends GeneralFunctionMap {
  static instance: Object;
  constructor() {
    // If the instance alread exists, then don't create a new instance.
    // If it doesn't, then setup some default values.
    if (!GlobalFunctionMap.instance) {
      super();
      GlobalFunctionMap.instance = this;
      this.map = {};
    }
    return GlobalFunctionMap.instance;
  }
}

/**
 * Function Map
 *
 * In FigureOne {@link Recorder}, state is saved in stringified javascript
 * objects. When the state includes a function (like a trigger method in an
 * animation for example) then that function is referenced in the state object
 * as a unique string id.
 *
 * When a geometry is loaded, functions that will be captured in state objects
 * need to be added to a function map. Both {@link Figure} and
 * {@link FigureElement} have attached function maps as the property `fnMap`.
 * Therefore, the method is added to either the figure or element function map,
 * with an associated unique string id, and that string id is used when the
 * function is used. For example as defined callbacks, triggerAnimationStep and
 * customAnimationStep callbacks, or recorder slide definitions (`entryState`,
 * `steadyState`, `leaveState`, `exec`, `execDelta`).
 *
 * The funciton map has:
 * - A map of functions
 * - A link to a global map of functions (a singleton)
 * - The ability to execute a function within the map
 *
 * @property {FunctionMap} global global function map
 * @property {Object} map local function map where keys are unique identifiers
 * and values are the associated functions
 *
 * @example
 * // Add a console function to a FunctionMap and execute it with a parameter
 * figure.fnMap.add('toConsole', s => console.log(s));
 * figure.fnMap.exec('toConsole', 'hello');
 */
class FunctionMap extends GeneralFunctionMap {
  map: { [id: string]: {
    fn: Function,
  }}

  global: GlobalFunctionMap;

  constructor() {
    super();
    this.global = new GlobalFunctionMap();
  }

  /**
   * Execute function with arguments.
   *
   * @param {string | Function | null} idOrFn If `string`, then a function
   * in the local map or global map will be executed (local map takes
   * precedence). If `Function`, then the
   * function will be exectuted. If `null`, then nothing will happen.
   * @param {any} ...args arguments to pass to function.
   */
  exec(idOrFn: string | Function | null, ...args: any) {
    return this.execOnMaps(idOrFn, [this.global.map], ...args);
  }

  execOnMaps(idOrFn: string | Function | null, mapsIn: Array<Object>, ...args: any) {
    return super.execOnMaps(idOrFn, [...mapsIn, this.global.map], ...args);
  }

  /**
   * Add a function to the map.
   * @param {string} id unique identifier
   * @param {Function} fn function to add
   */
  add(id: string, fn: Function) {
    return super.add(id, fn);
  }
}

export { FunctionMap, GlobalFunctionMap };
