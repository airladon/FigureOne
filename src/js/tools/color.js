// @flow

import colorNames from './colorNames';  // official css color names
// eslint-disable-next-line import/no-cycle
import getCSSColors from './getCssColors';
import { round } from './math';
// Function that converts any rgb or rgba string to an array of rgba numbers
// between 0 and 1
function RGBToArray(color: string): Array<number> {
  // Reduce the rgb(a) string to just numbers
  let colString: string = color;
  colString = colString.replace(/.*\(/i, '');
  colString = colString.replace(/\)/i, '');
  const strArray = colString.split(',');
  // Go through each rgb(a) value and normalize to 1.0
  const value: Array<number> = strArray.map((x, index) => {
    if (index < 3) {
      return parseInt(x, 10) / 255.0;
    }
    return parseFloat(x);
  });

  // If an alpha value isn't included, then include it with default value 1.0
  if (value.length === 3) {
    value.push(1);
  }
  return value;
}


// Function that converts any hex color string to an array of rgba numbers
// between 0 and 1 (where alpha is always 1)
function HexToArray(color: string): Array<number> {
  let colHex: string = color.slice(1);
  if (colHex.length < 6) {
    colHex =
      `${colHex[0]}${colHex[0]}${colHex[1]}${colHex[1]}${colHex[2]}${colHex[2]}`;
  }
  const col: Array<number> = [
    parseInt(colHex.slice(0, 2), 16) / 255.0,
    parseInt(colHex.slice(2, 4), 16) / 255.0,
    parseInt(colHex.slice(4, 6), 16) / 255.0,
    1,
  ];
  return col;
}

function cssColorToArray(cssColorString: string): Array<number> | null {
  const oNames = colorNames();               // Official css color names
  let color = cssColorString.slice(0);

  // If the color is an official name, then replace it with the hex rgb
  // equivalent
  if (color in oNames) {
    color = oNames[color];
  }

  // colorValue is the rgba array of colors between 0 and 1
  let colorValue: Array<number> = [];

  // If color string starts with 'rgb' (and therefore also 'rgba')
  if (color.toLowerCase().startsWith('rgb')) {
    colorValue = RGBToArray(color);

  // If color string starts with '#' it is hex
  } else if (color.startsWith('#')) {
    colorValue = HexToArray(color);
  }

  // If the color value array is defined, then add it to the final
  // dictionary
  if (colorValue.length > 0) {
    return colorValue;
  }
  return null;
}

function colorArrayToRGBA(color: TypeColor) {
  return `rgba(${
    Math.floor(color[0] * 255)},${
    Math.floor(color[1] * 255)},${
    Math.floor(color[2] * 255)},${
    color[3]})`;
}

function colorArrayToRGB(color: TypeColor) {
  return `rgb(${
    Math.floor(color[0] * 255)},${
    Math.floor(color[1] * 255)},${
    Math.floor(color[2] * 255)})`;
}

function areColorsSame(color1: Array<number>, color2: Array<number>, precision: number = 5) {
  if (color1.length !== color2.length) {
    return false;
  }
  for (let i = 0; i < color1.length; i += 1) {
    if (round(color1[i], precision) !== round(color2[i], precision)) {
      return false;
    }
  }
  return true;
}

function areColorsWithinDelta(color1: Array<number>, color2: Array<number>, delta: number = 0.001) {
  if (color1.length !== color2.length) {
    return false;
  }
  for (let i = 0; i < color1.length; i += 1) {
    const dC = Math.abs(color1[i] - color2[i]);
    if (dC > delta) {
      return false;
    }
  }
  return true;
}

export {
  RGBToArray, HexToArray, cssColorToArray, colorArrayToRGB,
  colorArrayToRGBA, getCSSColors, areColorsSame, areColorsWithinDelta,
};
