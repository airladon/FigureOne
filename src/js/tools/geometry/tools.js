// @flow
import { Point } from './Point';

function pointsToNumbers2(points: Array<Point>) {
  const list = Array(points.length * 2);
  for (let i = 0; i < points.length; i += 1) {
    list[i * 2] = points[i].x;
    list[i * 2 + 1] = points[i].y;
  }
  return list;
}

function pointsToNumbers(points: Array<Point>, dimension: 2 | 3 = 3) {
  if (dimension === 2) {
    return pointsToNumbers2(points);
  }
  const list = Array(points.length * 3);
  for (let i = 0; i < points.length; i += 1) {
    list[i * 3] = points[i].x;
    list[i * 3 + 1] = points[i].y;
    list[i * 3 + 2] = points[i].z;
  }
  return list;
}

function numbersToPoints2(numbers: Array<number>) {
  const points = Array(numbers.length / 2);
  for (let i = 0; i < points.length; i += 1) {
    points[i] = new Point(numbers[i * 2], numbers[i * 2 + 1]);
  }
  return points;
}

function numbersToPoints(numbers: Array<number>, dimension: 2 | 3 = 3) {
  if (dimension === 2) {
    return numbersToPoints2(numbers);
  }
  const points = Array(numbers.length / 3);
  for (let i = 0; i < points.length; i += 1) {
    points[i] = new Point(
      numbers[i * 3],
      numbers[i * 3 + 1],
      numbers[i * 3 + 2],
    );
  }
  return points;
}

function toNumbers(
  pointsOrNumbers: Array<Point> | Array<Array<number>> | Array<number>,
  dimension: 2 | 3 = 3,
) {
  if (pointsOrNumbers.length === 0) {
    return [];
  }
  if (typeof pointsOrNumbers[0] === 'number') {
    return pointsOrNumbers;
  }
  if (Array.isArray(pointsOrNumbers[0])) {
    const out = [];
    for (let i = 0; i < pointsOrNumbers.length; i += 1) {
      out.push(...pointsOrNumbers);
    }
    return out;
  }
  return pointsToNumbers(pointsOrNumbers, dimension);
}

function toPoints(pointsOrNumbers: Array<Point> | Array<number>, dimension: 2 | 3 = 3) {
  if (pointsOrNumbers.length === 0) {
    return [];
  }
  if (typeof pointsOrNumbers[0] !== 'number') {
    return pointsOrNumbers;
  }
  return numbersToPoints(pointsOrNumbers, dimension);
}

export {
  pointsToNumbers2,
  pointsToNumbers,
  numbersToPoints,
  numbersToPoints2,
  toNumbers,
  toPoints,
};
