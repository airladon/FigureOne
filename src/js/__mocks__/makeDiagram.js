// import {
//   DiagramElementPrimitive,
//   DiagramElementCollection,
//   // AnimationPhase,
// } from '../Element';
import Diagram from '../diagram/Diagram';
import {
  Rect, getPoint, Point,
} from '../tools/g2';
import webgl from './WebGLInstanceMock';
import DrawContext2D from './DrawContext2DMock';
// import VertexPolygon from '../DrawingObjects/VertexObject/VertexPolygon';
import * as tools from '../tools/tools';
import { round } from '../tools/math';

tools.isTouchDevice = jest.fn();

jest.mock('../diagram/Gesture');
jest.mock('../diagram/webgl/webgl');
jest.mock('../diagram/DrawContext2D');
jest.mock('../diagram/DrawingObjects/HTMLObject/HTMLObject');

const generateRandomStringMock = jest.spyOn(tools, 'generateRandomString');
generateRandomStringMock.mockImplementation(() => '000000');

export default function makeDiagram(
  inputCanvas = new Rect(100, -300, 1000, 500),
  inputLimits = new Rect(-1, -1, 2, 2),
) {
  document.body.innerHTML =
    '<div id="c">'
    + '  <canvas class="figureone__gl" id="id_figureone__gl__low">'
    + '  </canvas>'
    + '  <canvas class="figureone__text" id="id_figureone__text__low">'
    + '  </canvas>'
    + '  <div class="figureone__html">'
    + '  </div>'
    + '  <canvas class="figureone__gl" id="id_figureone__gl__high">'
    + '  </canvas>'
    + '  <canvas class="figureone__text" id="id_figureone__text__high">'
    + '  </canvas>'
    + '</div>';
  // canvas = document.getElementById('c');
  const definition = {
    width: inputCanvas.width,
    height: inputCanvas.height,
    limits: inputLimits,
  };

  const canvasMock = {
    width: definition.width,
    height: definition.height,
    // offsetLeft: 100,
    left: inputCanvas.left,
    // offsetTop: 200,
    top: inputCanvas.top,
    // width: definition.width,
    // height: definition.height,
    offsetWidth: definition.width,
    offsetHeight: definition.height,
    scrollLeft: 0,
    scrollTop: 0,
    // eslint-disable-next-line arrow-body-style
    getBoundingClientRect: () => {
      return {
        left: inputCanvas.left,
        top: inputCanvas.top,
        width: definition.width,
        height: definition.height,
      };
    },
  };
  const htmlCanvasMock = {
    style: {
      fontsize: 1,
    },
    offsetWidth: 100,
    appendChild: () => {},
  };
  const { limits } = definition;
  const diagram = new Diagram({ htmlId: 'c', limits });
  diagram.webglLow = webgl;
  diagram.webglHigh = webgl;
  diagram.webgl = webgl;
  diagram.canvasLow = canvasMock;
  diagram.canvasHigh = canvasMock;
  diagram.htmlCanvas = htmlCanvasMock;
  diagram.isTouchDevice = false;
  diagram.draw2DLow = new DrawContext2D(definition.width, definition.height);
  diagram.draw2DHigh = new DrawContext2D(definition.width, definition.height);
  diagram.draw2D = diagram.draw2DLow;
  diagram.shapesLow = diagram.getShapes(false);
  diagram.shapesHigh = diagram.getShapes(true);
  diagram.shapes = diagram.shapesLow;
  diagram.equationLow = diagram.getEquations(false);
  diagram.equationHigh = diagram.getEquations(true);
  diagram.equation = diagram.equationLow;
  diagram.advancedLow = diagram.getObjects(false);
  diagram.advancedHigh = diagram.getObjects(true);
  diagram.advanced = diagram.advancedLow;
  diagram.setSpaceTransforms();
  // needed as the first element needs to be set with the space Transforms
  diagram.initElements();
  diagram.mock = {
    initialTime: 0,
    duration: 0,
    previousTouchPoint: new Point(0, 0),
    timersBeforeDraw: true,
    timeStep: (deltaTimeInSecondsIn, frameTimeIn = null) => {
      const { duration, initialTime } = diagram.mock;
      const deltaTimeInSeconds = round(deltaTimeInSecondsIn, 8);
      let frameTime = deltaTimeInSeconds;
      // let deltaTime = 0;
      if (frameTimeIn != null && frameTimeIn < deltaTimeInSeconds) {
        frameTime = frameTimeIn;
        // deltaTime = frameTime;
      }
      let deltaTime = frameTime;

      // let delta = Math.min(deltaTimeInSeconds, frameTime);
      let lastTime = 0;
      deltaTime = round(deltaTime, 8);
      while (deltaTime <= round(deltaTimeInSeconds, 8)) {
        // if (window.asdf) {
        //   console.log(duration, deltaTime + 0.00);
        // }
        const newNow = round((duration + deltaTime + initialTime) * 1000, 8);
        global.performance.now = () => newNow;
        if (diagram.mock.timersBeforeDraw) {
          jest.advanceTimersByTime(round((deltaTime - lastTime) * 1000, 8));
        }
        diagram.animateNextFrame();
        diagram.draw(round(newNow / 1000, 8));
        if (!diagram.mock.timersBeforeDraw) {
          jest.advanceTimersByTime(round((deltaTime - lastTime) * 1000, 8));
        }
        lastTime = deltaTime;
        if (deltaTime === deltaTimeInSeconds || frameTime === 0) {
          deltaTime += 1;
        } else if (deltaTime + frameTime > deltaTimeInSeconds) {
          deltaTime = deltaTimeInSeconds;
        } else {
          deltaTime += frameTime;
        }
        deltaTime = round(deltaTime, 8);
      }
      diagram.mock.duration += deltaTimeInSeconds;
    },
    touchDown: (diagramPosition) => {
      const p = getPoint(diagramPosition);
      const pixelPoint = p.transformBy(diagram.spaceTransforms.diagramToPixel.m());
      const clientPoint = diagram.pixelToClient(pixelPoint);
      diagram.touchDownHandler(clientPoint);
      diagram.mock.previousTouchPoint = clientPoint;
    },
    touchUp: () => {
      diagram.touchUpHandler();
    },
    touchMove: (diagramPosition) => {
      const p = getPoint(diagramPosition);
      const pixelPoint = p.transformBy(diagram.spaceTransforms.diagramToPixel.m());
      const clientPoint = diagram.pixelToClient(pixelPoint);
      diagram.touchMoveHandler(diagram.mock.previousTouchPoint, clientPoint);
      diagram.mock.previousTouchPoint = clientPoint;
    },
  };
  diagram.globalAnimation.reset();
  return diagram;
}

