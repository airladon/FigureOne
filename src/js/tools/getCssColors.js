// @flow
// eslint-disable-next-line import/no-cycle
import { cssColorToArray } from './color';
import { getDefinedCSSVariables } from './getCssVariables';

const baseColors = [
  'blue',
  'red',
  'yellow',
  'green',
  'cyan',
  'brown',
  'orange',
  'violet',
  'grey',
  'black',
  'white',
];

const shades = [
  'base',
  'lighter',
  'light',
  'dark',
  'darker',
  'higher',
  'high',
  'low',
  'lower',
];

const paletteColorNames = [];
baseColors.forEach((color) => {
  shades.forEach((shade) => {
    paletteColorNames.push(`--palette-${color}-${shade}`);
  });
});

const diagramColorNames = [
  '--diagram-background',
  '--diagram-primary',
  '--diagram-warning',
  '--diagram-safe',
  '--diagram-passive',
  '--diagram-construction1',
  '--diagram-construction2',
  '--diagram-construction3',
  '--diagram-construction4',
  '--diagram-construction5',
  '--diagram-construction6',
  '--diagram-construction7',
  '--diagram-construction8',
  '--diagram-construction9',
  '--diagram-disabled',
  '--diagram-disabledDark',
  '--diagram-disabledDarker',
  '--diagram-disabledDarkest',
  '--diagram-push',
  '--diagram-action',
  '--diagram-text-base',
  '--diagram-text-warning',
  '--diagram-text-plot',
  '--diagram-text-keyword',
  '--diagram-text-keyword2',
  '--diagram-text-latin',
  '--diagram-text-greek',
  '--diagram-text-english',
  '--diagram-text-note',
  '--diagram-qr-background',
];

export default function getCSSColors(customColorNames: Array<string> = []): Object {
  let colors: Object = {};
  let paletteColors: Object = {};
  let diagramColors: Object = {};
  const colorNames = customColorNames.map(name => `--colors-${name}`);

  const { body } = document;
  if (body) {
    colors = getDefinedCSSVariables(
      body, colorNames, '--colors-', false,
      // $FlowFixMe
      cssColorToArray,
    );

    paletteColors = getDefinedCSSVariables(
      body, paletteColorNames, '--palette-', false,
      // $FlowFixMe
      cssColorToArray,
    );

    diagramColors = getDefinedCSSVariables(
      body, diagramColorNames, '--diagram-', false,
      // $FlowFixMe
      cssColorToArray,
    );
  }

  colors.palette = paletteColors;
  colors.diagram = diagramColors;
  return colors;
}

