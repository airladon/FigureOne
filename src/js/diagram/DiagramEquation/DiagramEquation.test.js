// import {
//   DiagramElementPrimitive,
//   DiagramElementCollection,
//   // AnimationPhase,
// } from '../Element';
// import Diagram from '../Diagram';
import {
  Point, Rect,
} from '../../tools/g2';
import {
  round,
} from '../../tools/math';
// import webgl from '../../__mocks__/WebGLInstanceMock';
// import DrawContext2D from '../../__mocks__/DrawContext2DMock';
import * as tools from '../../tools/tools';
import makeDiagram from '../../__mocks__/makeDiagram';

tools.isTouchDevice = jest.fn();

jest.mock('../Gesture');
jest.mock('../webgl/webgl');
jest.mock('../DrawContext2D');
// jest.mock('../../tools/tools');

describe('Diagram Equations From Object', () => {
  let diagram;
  let collection;
  let color1;
  let ways;

  beforeEach(() => {
    diagram = makeDiagram();
    collection = diagram.shapes.collection();
    color1 = [0.95, 0, 0, 1];
    ways = {
      simple: () => {
        diagram.equation.addEquation(collection, 'eqn', {
          color: color1,
          elements: {
            a: 'a',
            b: 'b',
            c: 'c',
            _2: '2',
            v: { symbol: 'vinculum' },
          },
          forms: {
            0: ['a', 'b', 'c'],
            1: [{ frac: ['a', 'v', '_2'] }, 'c'],
          },
          formSeries: ['0', '1'],
        });
        collection._eqn.showForm('1');
      },
      allOptions: () => {
        diagram.equation.addEquation(collection, 'eqn', {
          color: color1,
          position: [1, 1],           // Points can be defined as arrays
          scale: 0.85,
          elements: {
            a: 'a',
            b: 'b',
            c: 'c',
            _2: '2',
            v: { symbol: 'vinculum' },
          },
          formDefaults: {
            alignment: {
              fixTo: { x: 2, y: 2 },    // Points can also be defined as objects
              xAlign: 'right',
              yAlign: 'top',
            },
          },
          forms: {
            0: ['a', 'b', 'c'],
            1: [{ frac: ['a', 'v', '_2'] }, 'c'],
          },
          formSeries: ['0', '1'],
        });
        collection._eqn.showForm('1');
      },
      navigator: () => {
        diagram.equation.addEquation(collection, 'eqn', {
          color: color1,
          elements: {
            a: 'a',
            b: 'b',
            c: 'c',
            _2: '2',
            v: { symbol: 'vinculum' },
          },
          forms: {
            0: ['a', 'b', 'c'],
            1: [{ frac: ['a', 'v', '_2'] }, 'c'],
          },
          formSeries: ['0', '1'],
        });
        diagram.equation.addNavigator(collection, 'nav', {
          equation: collection._eqn,
          offset: new Point(0.2, -0.5),
          navType: '2Line',
          xAlign: 'center',
          yAlign: 'middle',
          navTypeOptions: { arrows: true },
        });
        // spoof a the navigator drawingObject
        collection._nav._table.drawingObject.element = { style: { opacity: 1 } };
        collection._nav.showForm('1');
      },
      navigatorWithoutEquationSplit: () => {
        diagram.equation.addNavigator(collection, 'test', {
          equation: collection._eqn,
          offset: [0.2, -0.5],
          navType: '2Line',
          xAlign: 'center',
          yAlign: 'middle',
          navTypeOptions: { arrows: true },
        });
        collection._testEqn.addElements({
          a: 'a',
          b: 'b',
          c: 'c',
          _2: '2',
          v: { symbol: 'vinculum' },
        });
        collection._testEqn.addForms({
          0: ['a', 'b', 'c'],
          1: [{ frac: ['a', 'v', '_2'] }, 'c'],
        });
        collection._testEqn.setFormSeries(['0', '1']);
        collection._testNav._table.drawingObject.element = { style: { opacity: 1 } };
        collection._testNav.showForm('1');
      },
      navigatorWithoutEquationAllInOne: () => {
        diagram.equation.addNavigator(collection, 'test', {
          color: color1,
          elements: {
            a: 'a',
            b: 'b',
            c: 'c',
            _2: '2',
            v: { symbol: 'vinculum' },
          },
          forms: {
            0: ['a', 'b', 'c'],
            1: [{ frac: ['a', 'v', '_2'] }, 'c'],
          },
          formSeries: ['0', '1'],
          offset: new Point(0.2, -0.5),
          navType: '2Line',
          xAlign: 'center',
          yAlign: 'middle',
          navTypeOptions: { arrows: true },
        });
        collection._testNav._table.drawingObject.element = { style: { opacity: 1 } };
        collection._testNav.showForm('1');
      },
    };
  });
  test('Diagram instantiation', () => {
    expect(diagram.limits).toEqual(new Rect(-1, -1, 2, 2));
  });
  describe('Equation Creation', () => {
    test('Simple', () => {
      ways.simple();
      expect(collection).toHaveProperty('_eqn');
      expect(collection._eqn).toHaveProperty('_a');
      expect(collection._eqn).toHaveProperty('_b');
      expect(collection._eqn).toHaveProperty('_c');
      expect(collection._eqn).toHaveProperty('__2');
      expect(collection._eqn).toHaveProperty('_v');

      tools.cleanUIDs(collection._eqn);
      expect(round(collection._eqn._a.transform.mat)).toMatchSnapshot();
      expect(round(collection._eqn.__2.transform.mat)).toMatchSnapshot();
      expect(round(collection._eqn._c.transform.mat)).toMatchSnapshot();
      expect(round(collection._eqn._v.transform.mat)).toMatchSnapshot();
    });
    test('Equation with all options', () => {
      ways.allOptions();
      expect(collection).toHaveProperty('_eqn');
      expect(collection._eqn).toHaveProperty('_a');
      expect(collection._eqn).toHaveProperty('_b');
      expect(collection._eqn).toHaveProperty('_c');
      expect(collection._eqn).toHaveProperty('__2');
      expect(collection._eqn).toHaveProperty('_v');

      tools.cleanUIDs(collection._eqn);
      expect(round(collection._eqn._a.transform.mat)).toMatchSnapshot();
      expect(round(collection._eqn.__2.transform.mat)).toMatchSnapshot();
      expect(round(collection._eqn._c.transform.mat)).toMatchSnapshot();
      expect(round(collection._eqn._v.transform.mat)).toMatchSnapshot();
    });
    test('Navigator with all options', () => {
      ways.navigator();
      expect(collection).toHaveProperty('_eqn');
      expect(collection).toHaveProperty('_nav');
      expect(collection._eqn).toHaveProperty('_a');
      expect(collection._eqn).toHaveProperty('_b');
      expect(collection._eqn).toHaveProperty('_c');
      expect(collection._eqn).toHaveProperty('__2');
      expect(collection._eqn).toHaveProperty('_v');

      tools.cleanUIDs(collection._eqn);
      expect(round(collection._eqn._a.transform.mat)).toMatchSnapshot();
      expect(round(collection._eqn.__2.transform.mat)).toMatchSnapshot();
      expect(round(collection._eqn._c.transform.mat)).toMatchSnapshot();
      expect(round(collection._eqn._v.transform.mat)).toMatchSnapshot();
    });
    test('Navigator without equation Split', () => {
      ways.navigatorWithoutEquationSplit();
      expect(collection).toHaveProperty('_testEqn');
      expect(collection).toHaveProperty('_testNav');
      expect(collection._testEqn).toHaveProperty('_a');
      expect(collection._testEqn).toHaveProperty('_b');
      expect(collection._testEqn).toHaveProperty('_c');
      expect(collection._testEqn).toHaveProperty('__2');
      expect(collection._testEqn).toHaveProperty('_v');

      tools.cleanUIDs(collection._testEqn);
      expect(round(collection._testEqn._a.transform.mat)).toMatchSnapshot();
      expect(round(collection._testEqn.__2.transform.mat)).toMatchSnapshot();
      expect(round(collection._testEqn._c.transform.mat)).toMatchSnapshot();
      expect(round(collection._testEqn._v.transform.mat)).toMatchSnapshot();
    });
    test('Navigator without equation all in one object', () => {
      ways.navigatorWithoutEquationAllInOne();
      expect(collection).toHaveProperty('_testEqn');
      expect(collection).toHaveProperty('_testNav');
      expect(collection._testEqn).toHaveProperty('_a');
      expect(collection._testEqn).toHaveProperty('_b');
      expect(collection._testEqn).toHaveProperty('_c');
      expect(collection._testEqn).toHaveProperty('__2');
      expect(collection._testEqn).toHaveProperty('_v');

      tools.cleanUIDs(collection._testEqn);
      expect(round(collection._testEqn._a.transform.mat)).toMatchSnapshot();
      expect(round(collection._testEqn.__2.transform.mat)).toMatchSnapshot();
      expect(round(collection._testEqn._c.transform.mat)).toMatchSnapshot();
      expect(round(collection._testEqn._v.transform.mat)).toMatchSnapshot();
    });
  });
});
