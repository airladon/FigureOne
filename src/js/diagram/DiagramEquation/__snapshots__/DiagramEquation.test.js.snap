// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Diagram Equations From Object Equation Creation Equation with all options 1`] = `
Array [
  0.85,
  0,
  1.9269,
  0,
  0.85,
  1.96345,
  0,
  0,
  1,
]
`;

exports[`Diagram Equations From Object Equation Creation Equation with all options 2`] = `
Array [
  0.85,
  0,
  1.9269,
  0,
  0.85,
  1.82813,
  0,
  0,
  1,
]
`;

exports[`Diagram Equations From Object Equation Creation Equation with all options 3`] = `
Array [
  0.85,
  0,
  1.966,
  0,
  0.85,
  1.85873,
  0,
  0,
  1,
]
`;

exports[`Diagram Equations From Object Equation Creation Equation with all options 4`] = `
Array [
  0.0442,
  0,
  1.9218,
  0,
  0.0085,
  1.91823,
  0,
  0,
  1,
]
`;

exports[`Diagram Equations From Object Equation Creation Navigator with all options 1`] = `
Array [
  0.7,
  0,
  0.0042,
  0,
  0.7,
  0.08624,
  0,
  0,
  1,
]
`;

exports[`Diagram Equations From Object Equation Creation Navigator with all options 2`] = `
Array [
  0.7,
  0,
  0.0042,
  0,
  0.7,
  -0.0252,
  0,
  0,
  1,
]
`;

exports[`Diagram Equations From Object Equation Creation Navigator with all options 3`] = `
Array [
  0.7,
  0,
  0.0364,
  0,
  0.7,
  0,
  0,
  0,
  1,
]
`;

exports[`Diagram Equations From Object Equation Creation Navigator with all options 4`] = `
Array [
  0.0364,
  0,
  0,
  0,
  0.008,
  0.049,
  0,
  0,
  1,
]
`;

exports[`Diagram Equations From Object Equation Creation Navigator without equation Split 1`] = `
Array [
  0.7,
  0,
  0.0042,
  0,
  0.7,
  0.08624,
  0,
  0,
  1,
]
`;

exports[`Diagram Equations From Object Equation Creation Navigator without equation Split 2`] = `
Array [
  0.7,
  0,
  0.0042,
  0,
  0.7,
  -0.0252,
  0,
  0,
  1,
]
`;

exports[`Diagram Equations From Object Equation Creation Navigator without equation Split 3`] = `
Array [
  0.7,
  0,
  0.0364,
  0,
  0.7,
  0,
  0,
  0,
  1,
]
`;

exports[`Diagram Equations From Object Equation Creation Navigator without equation Split 4`] = `
Array [
  0.0364,
  0,
  0,
  0,
  0.008,
  0.049,
  0,
  0,
  1,
]
`;

exports[`Diagram Equations From Object Equation Creation Navigator without equation all in one object 1`] = `
Array [
  0.7,
  0,
  0.0042,
  0,
  0.7,
  0.08624,
  0,
  0,
  1,
]
`;

exports[`Diagram Equations From Object Equation Creation Navigator without equation all in one object 2`] = `
Array [
  0.7,
  0,
  0.0042,
  0,
  0.7,
  -0.0252,
  0,
  0,
  1,
]
`;

exports[`Diagram Equations From Object Equation Creation Navigator without equation all in one object 3`] = `
Array [
  0.7,
  0,
  0.0364,
  0,
  0.7,
  0,
  0,
  0,
  1,
]
`;

exports[`Diagram Equations From Object Equation Creation Navigator without equation all in one object 4`] = `
Array [
  0.0364,
  0,
  0,
  0,
  0.008,
  0.049,
  0,
  0,
  1,
]
`;

exports[`Diagram Equations From Object Equation Creation Simple 1`] = `
Array [
  0.7,
  0,
  0.0042,
  0,
  0.7,
  0.08624,
  0,
  0,
  1,
]
`;

exports[`Diagram Equations From Object Equation Creation Simple 2`] = `
Array [
  0.7,
  0,
  0.0042,
  0,
  0.7,
  -0.0252,
  0,
  0,
  1,
]
`;

exports[`Diagram Equations From Object Equation Creation Simple 3`] = `
Array [
  0.7,
  0,
  0.0364,
  0,
  0.7,
  0,
  0,
  0,
  1,
]
`;

exports[`Diagram Equations From Object Equation Creation Simple 4`] = `
Array [
  0.0364,
  0,
  0,
  0,
  0.008,
  0.049,
  0,
  0,
  1,
]
`;
